mlir/lib/Dialect/Rock/Transforms/RockPipeline.cpp

```diff
@@ -293,33 +298,96 @@ void createSchedule(SmallVector<rock::StageOp> &stages,
   for (int t = 0; t < ii; t++) {
     int iteration = 0;
 
-    // The following stages need to run in parallel
+    DenseMap<rock::StageOp, int> stageIter;
+
+    // The following stages will run in parallel, but each
+    // stage needs to start at the right iteration
     SmallVector<rock::StageOp> parallelStages;
     for (size_t j = t; j < stages.size(); j += ii) {
-      schedule.push_back({stages[j], iteration++});
+      stageIter[stages[j]] = iteration++;
+      parallelStages.push_back(stages[j]);
+    }
+
+    // This is the set of multi-buffers needed at this time slot
+    // to ensure that the stage can run in parallel without messing
+    // each other's buffers
+    DenseMap<rock::GpuAllocOp, int> thisMultiBuffers = multiBuffers;
+    for (auto [alloc, factor] : thisMultiBuffers) {
+      thisMultiBuffers[alloc] = 1;
+    }
+
+    // Optimization: if there is a RAW register dependency (addrspace(5)) swap
+    // the stages. In this way, we don't need multibuffers (i.e., we read the
+    // buffer first and then we write into it). From the point of view of the
+    // stages, they don't care because they belong to different iterations. In
+    // theory this could be applied to any buffer, but for proper memory, this
+    // can be more expensive (i.e., you need barriers)
+    DenseMap<int, SmallVector<int>> swapCandidates;
```

Comment:
Vague nit: might be a decent idea to go for `unsigned` here? Using `unsigned` for parameter indexing and the like seems pretty common in LLVM code. But that's not critical.